yt_dlp/extractor/epoch.py

```diff
@@ -35,21 +49,59 @@ class EpochIE(InfoExtractor):
                 'id': '9489f994-2a20-4812-b233-ac0e5c345632',
                 'ext': 'mp4',
                 'title': 'Dick Morris Discusses His Book ‘The Return: Trump’s Big 2024 Comeback’',
+                'description': 'md5:ef6f45dd925b5d23e2a862487b42030e',
+                'thumbnail': r're:https://.*\.jpg',
             }
         },
+        {
+            'note': 'Georestricted to US',
+            'url': 'https://www.theepochtimes.com/silence-patton-documentary_4849547.html',
+            'info_dict': {
+                'id': '4849547',
+                'ext': 'mp4',
+                'title': 'Silence Patton | Documentary',
+                'description': 'Why was General Patton silenced during his service in World War II?',
+                'thumbnail': r're:https://.*\.jpg',
+            },
+            'params': {'skip_download': True},
+            'expected_warnings': [
+                'This film is only available in the United States because of territorial licensing.'
+            ],
+        },
     ]
 
     def _real_extract(self, url):
         video_id = self._match_id(url)
         webpage = self._download_webpage(url, video_id)
+        # get_element_by_class() fails on some websites, let's do it the dirty way ...
```

Comment:
Example? I would prefer to improve the function instead

Reply:
https://www.theepochtimes.com/2021-epoch-experience-package-the-7-churches-of-revelation-times-of-fire_4112140.html
https://www.theepochtimes.com/the-great-eclipse-the-creation-guys-series_4061624.html

```
Traceback (most recent call last):
  File yt-dlp/yt_dlp/YoutubeDL.py", line 1493, in wrapper
    return func(self, *args, **kwargs)
  File yt-dlp/yt_dlp/YoutubeDL.py", line 1569, in __extract_info
    ie_result = ie.extract(url)
  File yt-dlp/yt_dlp/extractor/common.py", line 674, in extract
    ie_result = self._real_extract(url)
  File yt-dlp/yt_dlp/extractor/epoch.py", line 78, in _real_extract
    raw_description = clean_html(get_element_by_class('article_content', webpage))
  File yt-dlp/yt_dlp/utils.py", line 372, in get_element_by_class
    retval = get_elements_by_class(class_name, html)
  File yt-dlp/yt_dlp/utils.py", line 394, in get_elements_by_class
    return get_elements_by_attribute(
  File yt-dlp/yt_dlp/utils.py", line 408, in get_elements_by_attribute
    return [content for content, _ in get_elements_text_and_html_by_attribute(*args, **kwargs)]
  File yt-dlp/yt_dlp/utils.py", line 408, in <listcomp>
    return [content for content, _ in get_elements_text_and_html_by_attribute(*args, **kwargs)]
  File yt-dlp/yt_dlp/utils.py", line 435, in get_elements_text_and_html_by_attribute
    content, whole = get_element_text_and_html_by_tag(m.group('tag'), html[m.start():])
  File yt-dlp/yt_dlp/utils.py", line 513, in get_element_text_and_html_by_tag
    parser.feed(html[offset:offset + next_closing_tag_end])
  File "python3.10/html/parser.py", line 110, in feed
    self.goahead(0)
  File "python3.10/html/parser.py", line 172, in goahead
    k = self.parse_endtag(i)
  File "python3.10/html/parser.py", line 420, in parse_endtag
    self.handle_endtag(elem)
  File yt-dlp/yt_dlp/utils.py", line 480, in handle_endtag
    raise compat_HTMLParseError(f'matching opening tag for closing {tag} tag not found')
yt_dlp.compat.compat_HTMLParseError: matching opening tag for closing figcaption tag not found
```

Should I use it anyway?

Reply:
I checked and the html is indeed malnested. Should we be less strict and just removing the tag from tagstack instead of trimming?

```patch
diff --git a/yt_dlp/utils.py b/yt_dlp/utils.py
--- a/yt_dlp/utils.py	
+++ b/yt_dlp/utils.py	
@@ -467,16 +467,14 @@
         pass
 
     def handle_starttag(self, tag, _):
-        self.tagstack.append(tag)
+        self.tagstack.appendleft(tag)
 
     def handle_endtag(self, tag):
         if not self.tagstack:
             raise compat_HTMLParseError('no tags in the stack')
-        while self.tagstack:
-            inner_tag = self.tagstack.pop()
-            if inner_tag == tag:
-                break
-        else:
+        try:
+            self.tagstack.remove(tag)
+        except ValueError:
             raise compat_HTMLParseError(f'matching opening tag for closing {tag} tag not found')
         if not self.tagstack:
             raise self.HTMLBreakOnClosingTagException()
```

Reply:
Yes. The parser is way too strict for what we need it to do. We should also add a corresponding test for `get_element_text_and_html_by_tag`

Reply:
I already played around with the code and can open a PR...